File: cpp/data_structures/tree/trie/trie.cpp
#include <bits/stdc++.h>
using namespace std;

struct TrieNode {
    array<TrieNode*, 26> ch{};
    bool end = false;
};

struct Trie {
    TrieNode* root = new TrieNode();

    void insert(const string& w) {
        TrieNode* n = root;
        for (char c : w) {
            int i = c - 'a';
            if (!n->ch[i]) n->ch[i] = new TrieNode();
            n = n->ch[i];
        }
        n->end = true;
    }

    bool search(const string& w) const {
        TrieNode* n = root;
        for (char c : w) {
            int i = c - 'a';
            if (!n->ch[i]) return false;
            n = n->ch[i];
        }
        return n->end;
    }

    bool startsWith(const string& p) const {
        TrieNode* n = root;
        for (char c : p) {
            int i = c - 'a';
            if (!n->ch[i]) return false;
            n = n->ch[i];
        }
        return true;
    }
};

int main() {
    Trie t;
    t.insert("apple");
    assert(t.search("apple") && !t.search("app"));
    assert(t.startsWith("app"));
    cout << "Trie OK" << endl;
    return 0;
}

File: python/data_structures/tree/trie/trie.py
"""Trie：小写字母 a-z。"""

from __future__ import annotations


class Trie:
    def __init__(self) -> None:
        self._root: dict[str, dict] = {}

    def insert(self, word: str) -> None:
        cur: dict[str, dict] = self._root
        for ch in word:
            cur = cur.setdefault(ch, {})
        cur["$"] = {}  # end marker

    def search(self, word: str) -> bool:
        cur = self._root
        for ch in word:
            if ch not in cur:
                return False
            cur = cur[ch]  # type: ignore[assignment]
        return "$" in cur

    def starts_with(self, prefix: str) -> bool:
        cur = self._root
        for ch in prefix:
            if ch not in cur:
                return False
            cur = cur[ch]  # type: ignore[assignment]
        return True


if __name__ == "__main__":
    t = Trie()
    t.insert("apple")
    assert t.search("apple") and not t.search("app")
    assert t.starts_with("app")
    t.insert("")
    assert t.search("") and t.starts_with("")
    print("Trie OK")
